scraper.py

```diff
@@ -22,30 +24,123 @@
+    return BeautifulSoup(html, 'html.parser', from_encoding=encoding)
+
+
+def extract_data_listings(html):
+    pattern = re.compile(r'PR[\d]+~')
+    return html.find_all('div', id=pattern)
+
+
+def has_two_tds(element):
+    is_row = element.name == 'tr'
+    children = element.find_all('td', recursive=False)
+    has_two_children = len(children) == 2
+    return is_row and has_two_children
+
+
+def clean_data(data):
+    data = data.string
+    try:
+        return data.strip(' \n:-')
+    except AttributeError:
+        return ''
+
+
+def extract_restaurant_metadata(element):
+    metadata_rows = element.find('table').find_all(has_two_tds,
+                                                   recursive=False)
+    rdata = {}
+    current_label = ''
+    for row in metadata_rows:
+        key_cell, value_cell = row.find_all('td', recursive=False)
+        new_label = clean_data(key_cell)
+        if new_label:
+            current_label = new_label
+        rdata.setdefault(current_label, []).append(clean_data(value_cell))
+    return rdata
+
+
+def is_inspection_row(element):
+    is_row = element.name == 'tr'
+    if not is_row:
+        return False
+    cells = element.find_all('td', recursive=False)
+    has_four = len(cells) == 4
+    this_text = clean_data(cells[0]).lower()
+    contains_word = 'inspection' in this_text
+    does_not_start = not this_text.startswith('inspection')
+    return is_row and has_four and contains_word and does_not_start
+
+
+def extract_score_data(element):
+    inspection_rows = element.find_all(is_inspection_row)
+    samples = len(inspection_rows)
+    total = high_score = average = 0
+    for row in inspection_rows:
+        string_value = clean_data(row.find_all('td')[2])
+        try:
+            int_value = int(string_value)
+        except (ValueError, TypeError):
+            samples -= 1
+        else:
+            total += int_value
+            high_score = int_value if int_value > high_score else high_score
+    if samples:
+        average = total/float(samples)
+    data = {
+        u'Average Score': average,
+        u'High Score': high_score,
+        u'Total Inspections': samples
+    }
+    return data
 
 
-# with io.open('kc_health_data.html') as f:
-#                html_doc = f.read()
-#
-# soup = BeautifulSoup(html_doc, 'html.parser')
-#
-# print(soup)
+if __name__ == '__main__':
+    kwargs = {
+        'Inspection_Start': '2/1/2013',
+        'Inspection_End': '2/1/2015',
+        'Zip_Code': '98109'
+    }
+    if len(sys.argv) > 1 and sys.argv[1] == 'test':
+        source, encoding = load_inspection_page()
+    else:
+        source, encoding = get_inspection_page(**kwargs)
+    doc = parse_source(source, encoding)
+    listings = extract_data_listings(doc)
+    for listing in listings:
+        metadata = extract_restaurant_metadata(listing)
+        score_data = extract_score_data(listing)
+        print({'metadata': metadata, 'score': score_data})
```

Comment:
You can also update the metadata with the score data to end up with one dictionary for each restaurant that has all the information you need:
```python
metadata = metadata.update(extract_score_data(listing))
```